deploy/sample-backup-configs/sample_config_build.ini

```diff
@@ -1,6 +1,21 @@
 # BUILDTIME/AGFI management configuration for the FireSim Simulation Manager
 # See docs/Advanced-Usage/Manager/Manager-Configuration-Files.rst for documentation of all of these params.
 
+[buildfarmconfig]
+# build platform (options are "aws" or "local")
+hostplatform=aws
```

Comment:
What if I want to run the `replace-rtl` and `aws-fpga` Vivdo build flow locally and then submit my checkpoint to S3 for an AMI to be created?  I've been asked whether this is possible.  We should think through what that would look like so that choices we make in the config file here don't end up being really confusing if/when that is added.

Reply:
> > > Want from Sagar, have local build support on AWS so that you don't have to spawn a buildfarm instance.
> > 
> > 
> > How would you expect it to behave if you have multiple configurations? The easy thing to do right now is run them all in parallel on the localhost. Maybe it's also easy to make them sequential but it would take me more looking into how people have `@parallel` and `@sequential` variants of the same task with Fabric.
> 
> I haven't thought in-depth on this... Sagar complained about it to me offline so I wanted it to be documented somewhere. My assumption if that things are done locally things should be done sequentially (IIRC Vivado uses 8-ish threads... so I don't thing running 8+ builds in parallel is recommended).

I made my comment before reading your code and I think we should maybe move this thread of conversation here so that the main unthreaded conversation in this PR doesn't become a ton of quoted replies that keep getting longer and longer... 

After reading through this part of the code and thinking about this more, it seems like we have two dimensions here:
1. buildhosts - aws, local (maybe there will need to be different kinds of local)
    * this could be as simple as aws provisions the buildcluster for you while local just uses a list of hostnames you provide
    * in either case, if you have fewer hostnames than configs you want to build, we could start by failing like you currently do. 
2. fpgaplatform - AWS.f1, other boards, is this not already provided by `PLATFORM_CONFIG` in the individual `build_config_recipes.ini`?  Would it be better if we just provided a mapping from values of scala `PLATFORM_CONFIG` to the task implementation class that needs to be used in python?  Perhaps longer term, the `PLATFORM_CONFIG` should be factored out into a JSON or YAML or something that both the Scala and Python could refer to for a single source of truth about the platform?  Or maybe the Scala PlatformConfig _generates_ Python or YAML or something that the firesim manager reads out of 

The existing firesim manager knows how to do `buildhosts=aws fpgaplatform=f1`.  In this PR, you're trying to teach it to do `buildhosts=local fpgaplatform=xilinx-u250`.  Sagar complained about wanting to be able to do `buildhosts=local fpgaplatform=f1`.

If we view it like this, the config file might be better like:
```
[buildhosts]
user@host1.mydatacenter
user@host2.mydatacenter
david@davidslaptop.net #because he wants to run my builds
somecustomthing
aws

[hostproviders]
aws=ProvisionEC2BuildFarm  #ProvisionEC2BuildFarm does the spinup and teardown of hosts that exists now
somecustomthing=NonBARPluginClass 
```


The way this would work is that the buildhosts is an ordered list that gets zipped with [builds] to create the mapping of  buildhost to build.  However, there's one fancy thing about iterating the [buildhosts], if you hit the name of a buildhost that is defined in [hostproviders], the provided class name acts like a factory that provisions a host for you and returns the IP address you use to access it.  I'm envisioning that we might also want to have the hostproviders be albe to return something that says "sorry bruh, can't give you any more hosts".  This would be useful in the case where you have a local cluster that can provide some number of hosts but then you have the aws hostprovider at the end and it can burst up to whatever your limit for EC2 hosts is (but even then it should probably learn that it's unable to provision more hosts and quit).  

Perhaps what I just described is that we want to zip [builds] with a flatMap of buildhosts where the hostproviders themselves are iterable.

I can see how this same idea could be applied at runtime (provided that we change things a little in config_hwdb.ini so that we could specify the target we want to simulate and the firesim manager learns how to pick the appropriate PLATFORM_CONFIG based on the host it is given to run on), maybe PLATFORM_CONFIG becomes a dict of configs to bitstream identifiers or something.

These are just thoughts.

Reply:
> What if I want to run the `replace-rtl` and `aws-fpga` Vivdo build flow locally and then submit my checkpoint to S3 for an AMI to be created? I've been asked whether this is possible. We should think through what that would look like so that choices we make in the config file here don't end up being really confusing if/when that is added.

I think this should just work if you have the AWS CLI set up? At the very least, it shouldn't be too hard.

Reply:
I like your orthogonalization of concerns. We should move it to a separate issue though. Run & BuildFarm provisioning does seem similar, we could probably pull the hostproviders stuff + fixed hosts in a single `host_config.ini`. If we factored it nicely it would be pretty easy to write a slurm / qsub module that could dynamically provide hosts as we do in AWS.

Reply:
> We should move it to a separate issue though.
Sounds good to me.